discover.go

```diff
@@ -48,10 +55,48 @@ type Discover struct {
 	// If nil, the default list of providers is used.
 	Providers map[string]Provider
 
+	// userAgent is the string to use for requests, when supported.
+	userAgent string
+
 	// once is used to initialize the actual list of providers.
 	once sync.Once
 }
 
+// Option is used as an initialization option/
+type Option func(*Discover) error
+
+// NewDiscover creates a new discover client with the given options.
+func NewDiscover(opts ...Option) (*Discover, error) {
```

Comment:
If I'm reading this right you're setting up a new way to initialize go-discover. Is it possible to limit the changes to fit with the existing APIs? If not, can you describe the use-case and example in Consul (I believe you're planning on utilizing this over there, per your comment)?

Reply:
Also I'd name this `New`, not `NewDiscover`, since the package is already `discover`. This is just some guideline from "writing effective go".

Reply:
Renamed to `New`. Thanks for the feedback.

The existing API doesn't change and is not obsoleted. However, I didn't want to make `UserAgent` a publicly accessible field on the Discover struct because it would introduce race conditions.

The API current is:

```go
d := discover.Discover{}
```

That API will continue to function, but it's not very flexible, so I adopted the "options pattern" to allow for this:

```go
d := discover.New(discovery.WithUserAgent("consul/1.0"))
```

Reply:
To be clearer: there exists a sad race condition in the current implementation. Since `Providers` is a public field on the struct, a user can set `Providers` to something, call a command, and then alter the field. However, since the init is done in a `once`, those changes won't take place.